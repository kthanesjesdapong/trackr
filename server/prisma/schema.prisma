// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

generator client {
  provider = "prisma-client-js"
}

generator pothos {
  provider = "prisma-pothos-types"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model Problem {
  id             Int       @id
  frontendId     Int       
  title          String
  titleSlug      String
  difficulty     String
  acRate         Float
  topics TopicOnProblems[]
  topicTags Json?
  //A problemAttempt can have zero or more attempts
  problemAttempts ProblemAttempt[]

  @@unique([id,title])
}

model Topic {
  id             Int      @id @default(autoincrement())
  topicSlug      String   @unique
  topicName      String   @unique
  problems TopicOnProblems[]

  @@unique([id,topicSlug,topicName])
}


model TopicOnProblems {
  problem   Problem   @relation(fields:[problemId], references: [id])
  topic     Topic     @relation(fields:[topicId,topicName,topicSlug], references:[id, topicName, topicSlug])

  problemId Int

  topicId    Int
  topicSlug String
  topicName String
  @@id([problemId, topicId])
}

//Add more fields for User Table Later // PW, EMAIL
model User {
  id String @id @default(uuid())
  registeredAt DateTime @default(now())
  //A user can have zero or more problemAttempts
  problemAttempts ProblemAttemptDetails[]
}


//Each row within ProblemAttempt will be an attempt
model ProblemAttempt {
  id             Int      @id @default(autoincrement())
  createdAt      DateTime @default(now())
  backendId     Int       @default(autoincrement())

  problem        Problem @relation(fields:[problemId], references: [id])
  // A problem attempt must always have a problemId
  problemId      Int


  // ProblemAttemptDetails
  problemAttemptDetail ProblemAttemptDetails @relation(fields:[attemptDetailId], references: [id])
  //attemptDetailId must be unique
  attemptDetailId Int @unique

}

model ProblemAttemptDetails {
  id        Int      @id @default(autoincrement())

  //A problemAttemptDetails must have one user
  user      User     @relation(fields:[userId], references:[id])
  userId    String

  //ProblemAttemptDetails may have zero or one ProblemAttemptId
  ProblemAttempt ProblemAttempt?
  status    String
  createdAt DateTime @default(now())
}
