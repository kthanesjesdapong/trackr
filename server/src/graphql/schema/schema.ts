import { builder } from '@graphql/schema/builder';
import '@graphql/schema/index';

//these codes are being importing to builder
export const schema = builder.toSchema({});

