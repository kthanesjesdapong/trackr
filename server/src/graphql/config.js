// ./pothos.config.js

/** @type {import('prisma-generator-pothos-codegen').Config} */
module.exports = {
    inputs: {
        outputFilePath: './src/graphql/__generated__/inputs.ts',
    },
    crud: {
        outputDir: './src/graphql/__generated__/',
        inputsImporter: `export * as Inputs from '@graphql/__generated__/inputs';`,
        resolversImports: `import prisma from '@lib/prisma';`,
        prismaCaller: 'prisma',
    },
    global: {
        builderImporter: `import { builder } from '@graphql/builder';`,
    },
};