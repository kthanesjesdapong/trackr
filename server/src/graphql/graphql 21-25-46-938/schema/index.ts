export * from './models/problems/Problem';
export * from './models/topics/Topic';
export * from './models/problems/TopicOnProblems';
export * from './models/users/User';
export * from './models/problemAttempt/ProblemAttempt';
export * from './models/problemAttemptDetails/ProblemAttemptDetail';
export * from './resolvers/query/Query';
export * from './resolvers/mutation/Mutation';