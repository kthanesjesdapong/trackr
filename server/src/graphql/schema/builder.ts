import SchemaBuilder from '@pothos/core';
import { DateResolver } from 'graphql-scalars';
import PrismaPlugin from '@pothos/plugin-prisma';
import type PrismaTypes from '@pothos/plugin-prisma/generated';
import { prisma } from '../../prisma/client';

//Setting up genereric for custom scalars
export const builder = new SchemaBuilder<{
  Scalars: {
    Date: { Input: Date; Output: Date; };
  };
  PrismaTypes: PrismaTypes;
}>({
  plugins: [PrismaPlugin],
  prisma: {
    client: prisma
  }
});

//init queryType
builder.queryType({});

//gQL doesnt have DATE type.
builder.addScalarType("Date", DateResolver, {});
