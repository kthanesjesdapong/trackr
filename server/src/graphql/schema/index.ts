export * from './typedefs/problems/Problem';
export * from './typedefs/topics/Topic';
export * from './typedefs/problems/TopicOnProblems';
export * from './typedefs/users/User';
export * from './resolvers/query/Query';