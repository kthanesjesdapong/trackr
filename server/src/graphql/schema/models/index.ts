export * from '@graphql/schema/models/problems/Problem';
export * from '@graphql/schema/models/topics/Topic';
export * from '@graphql/schema/models/problems/TopicOnProblems';
export * from '@graphql/schema/models/users/User';
export * from '@graphql/schema/models/problemAttempt/ProblemAttempt';
export * from '@graphql/schema/models/problemAttemptDetails/ProblemAttemptDetail'




