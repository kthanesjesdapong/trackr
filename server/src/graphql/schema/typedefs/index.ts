export * from './Problem';
export * from './Topic';
export * from './Query';
export * from './TopicOnProblems';