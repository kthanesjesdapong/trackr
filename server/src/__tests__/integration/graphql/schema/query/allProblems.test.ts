//Import schema
import '../../../../../graphql/schema/index';
import { execute } from 'graphql';
import { gql } from 'graphql-tag';
import pothosMock from '../../../../__mocks__/pothosMock';
import prismaMock from '../../../../__mocks__/prismaMock';
import { builder } from '../../../../../graphql/schema/builder';


describe('mock', () => {
  it('queries all problems', async () => {
    const mockedSchema = await builder.toSchema({
      mocks: {
        User: {
          id: () => prismaMock.user.findMany
          // id: () =>
          // title: () => 
        }
      }
    });

    const query = gql`
    mutation {
      createUser(id: "kavin1234567"){
        id
      }
    }`;

    const result = await execute({
      schema: mockedSchema,
      document: query,
      contextValue: {}
    });

    expect(result).toMatchSnapshot();
    // const user = await prismaMock.user.create({
    //   data: {
    //     id: 'kavin1234567',
    //     registeredAt: new Date()
    //   }
    // });


  });
});


