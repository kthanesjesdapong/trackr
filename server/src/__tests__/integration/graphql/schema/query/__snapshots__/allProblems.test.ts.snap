// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mock queries all problems 1`] = `
{
  "data": null,
  "errors": [
    [GraphQLError: Cannot read properties of undefined (reading 'user')],
  ],
}
`;
